File: src/CSharp/Benchmarking.Common/HardwareInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using Medallion.Shell;

namespace Benchmarking.Common
{
    public static class HardwareInfo
    {
        public static string GetCpuModelName()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) {
                var cmd = Command.Run("wmic", "cpu", "get", "name");
                return cmd.StandardOutput.GetLines().SkipEmpty().Last().Trim();
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) {
                var cmd = Command.Run("cat", "/proc/cpuinfo");
                return cmd.StandardOutput.GetLines().SkipEmpty()
                    .ToPairs().First(p => p.Name == "model name").Value;
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) {
                var cmd = Command.Run("sysctl", "-n", "machdep.cpu.brand_string");
                return cmd.StandardOutput.GetLines().SkipEmpty().Last().Trim();
            }
            else 
                throw new NotSupportedException($"Unsupported OS: {RuntimeInformation.OSDescription}");
        }

        // In GB
        public static int? GetRamSize()
        {
            try {
                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) {
                    var cmd = Command.Run("wmic", "computersystem", "get", "TotalPhysicalMemory");
                    var stringValue = cmd.StandardOutput.GetLines().SkipEmpty().Last().Trim();
                    return (int) Math.Round(long.Parse(stringValue) / Sizes.GB);
                }
                if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) {
                    var cmd = Command.Run("cat", "/proc/meminfo");
                    var stringValue = cmd.StandardOutput.GetLines().SkipEmpty()
                        .ToPairs().Single(p => p.Name == "MemTotal")
                        .Value.Split(' ', StringSplitOptions.RemoveEmptyEntries).First();
                    return (int) Math.Round(long.Parse(stringValue) * Sizes.KB / Sizes.GB);
                }
                if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) {
                    var cmd = Command.Run("system_profiler", "SPHardwareDataType");
                    var stringValue = cmd.StandardOutput.GetLines().SkipEmpty()
                        .ToPairs().Single(p => p.Name == "Memory")
                        .Value.Split(' ', StringSplitOptions.RemoveEmptyEntries).First();
                    return (int) Math.Round(double.Parse(stringValue));
                }
                return null;
            }
            catch {
                return null;
            }
        }
        
        private static IEnumerable<string> SkipEmpty(this IEnumerable<string> lines) =>
            lines.Select(l => l.TrimEnd()).Where(l => !string.IsNullOrEmpty(l));

        private static IEnumerable<(string Name, string Value)> ToPairs(
            this IEnumerable<string> lines, string separator = ":")
        {
            foreach (var line in lines) {
                var i = line.IndexOf(separator, StringComparison.Ordinal);
                if (i > 0)
                    yield return (line.Substring(0, i).Trim(), line.Substring(i + 1).Trim()); 
            }
        }
    }
}

File: src/Go/common/hardware_info.go
package common

import (
	"errors"
	"fmt"
	"log"
	"math"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
)

func GetOSVersion() string {
	switch runtime.GOOS {
	case "windows":
		return runAndGetLastLine("ver")
	case "darwin": // Mac OS X, iOS
		return runAndGetLastLine("system_profiler SPSoftwareDataType | awk '( $1 == \"System Version:\" ) { print $2 }'")
	default:
		return runAndGetLastLine("cat /proc/version")
	}
}

func GetCpuModelName() string {
	switch runtime.GOOS {
	case "windows":
		return runAndGetLastLine("wmic cpu get name")
	case "darwin": // Mac OS X, iOS
		return runAndGetLastLine("sysctl -n machdep.cpu.brand_string")
	default:
		return runAndGetLastLine("lscpu | grep 'Model name:' | sed --expression='s/.*:\\s*//g'")
	}
}

func GetRamSize() int {
	switch runtime.GOOS {
	case "windows":
		sizeB := runAndParseFloat("wmic computersystem get TotalPhysicalMemory")
		return int(math.Round(sizeB / GB))
	case "darwin": // Mac OS X, iOS
		sizeGb := runAndParseFloat("system_profiler SPHardwareDataType | awk '( $1 == \"Memory:\" ) { print $2 }'")
		return int(math.Round(sizeGb))
	default:
		sizeKb := runAndParseFloat("cat /proc/meminfo | awk '( $1 == \"MemTotal:\" ) { print $2 }'")
		return int(math.Round(sizeKb * KB / GB))
	}
}

func runAndParseFloat(cmd string) float64 {
	output := runAndGetLastLine(cmd)
	result, err := strconv.ParseFloat(output, 0)
	if err != nil {
		log.Fatal(err)
	}
	return result
}

func runAndGetLastLine(cmd string) string {
	output := run(cmd)
	output = strings.Replace(output, "\r\n", "\n", -1) // Windows output handling
	lines := strings.Split(output, "\n")
	for i := len(lines) - 1; i >= 0; i-- {
		r := strings.Trim(lines[i], " \t\r\n")
		if r != "" {
			return r
		}
	}
	return ""
}

func run(cmd string) string {
	command := "bash"
	arg := "-c"
	if runtime.GOOS == "windows" {
		command = "cmd"
		arg = "/C"
	}
	out, err := exec.Command(command, arg, cmd).Output()
	if err != nil {
		log.Fatal(err)
	}
	return string(out)
}

func failUnsupportedOS() {
	goos := runtime.GOOS
	err := errors.New(fmt.Sprintf("Unsupported OS: %v", goos))
	log.Fatal(err)
}
